### DISCUSSION TITLE
Prepared statements - psycopg3, psycopg2

### DISCUSSION THREAD

**Discussion by joseortiz3** [pgbouncer/pgbouncer#995] [Q&A] [answered]
The current [psycopg3 documentation](https://www.psycopg.org/psycopg3/docs/advanced/prepare.html) says:

> Note however that Psycopg doesn’t use SQL statements such as PREPARE and EXECUTE, but protocol level commands such as the ones exposed by [PQsendPrepare](https://www.postgresql.org/docs/14/libpq-async.html#LIBPQ-PQSENDPREPARE), [PQsendQueryPrepared](https://www.postgresql.org/docs/14/libpq-async.html#LIBPQ-PQSENDQUERYPREPARED). 
> Using external connection poolers, such as PgBouncer, is not compatible with prepared statements, because the same client connection may change the server session it refers to.

The current [pgbouncer documentation](https://www.pgbouncer.org/config.html#max_prepared_statements) says:

> Note: This tracking and rewriting of prepared statement commands does not work for SQL-level prepared statement commands such as PREPARE, EXECUTE, DEALLOCATE, DEALLOCATE ALL and DISCARD ALL. Running DEALLOCATE ALL and DISCARD ALL is especially problematic, since those commands appear to run successfully, but they mess up with the state of the server connection significantly without PgBouncer noticing. Which in turn will very likely break the execution of any further prepared statements on that server connection.

Is the psycopg3 documentation outdated now? Is pgbouncer known to work with it?

Also, I take it, psycopg2 doesn't do protocol-level prepared statements at all, correct?

**Comment by JelteF** [comment, marked as answer]
Psycopg3 docs are outdated now. However, it's not completely working with PgBouncer at the moment. The reason is that to close prepared statements it uses the DEALLOCATE sql command. 

You can disable the deallocation behaviour in psycopg by setting `prepare_threshold` to `float('inf')` (or some very high number).

Related links:
- Tracking issue at psycopg: https://github.com/psycopg/psycopg/issues/589
- Similar issue with pdo (PHP): https://github.com/pgbouncer/pgbouncer/issues/991
- Prerequisite to fix this issue that I contributed to postgres: https://github.com/postgres/postgres/commit/28b5726561841556dc3e00ffe26b01a8107ee654
- Prerequisite to fix this issue that I contributed to psycopg: https://github.com/psycopg/psycopg/pull/603

**Reply by riteshgl** [reply]
@JelteF we are facing a similar issue in our project based on RoR. 
In Ruby ActiveRecord also a DEALLOCATE sql command is being used. Setting the statement_limit for our service to a very high number is one the of the solutions but one thing I wanted to understand is in this case what should be the value of max_prepared_statements on pgBouncer? 
Should it be higher than the limit set at RoR service level or will setting it to a lower value also work fine?

**Reply by JelteF** [reply]
Setting `max_prepared_statements` to a lower value on PgBouncer will work fine, and even is the recommended approach. Otherwise the connections to Postgres will probably have many prepared statements (which can result in perf issues at the Postgres side).

**Reply by riteshgl** [reply]
@JelteF Thanks for the response. I see that setting the max_prepared_statement count lower works as pgBouncer stores all the prepared statements in a global cache and uses global cache to get the prepared statement. 
I wanted to know a few more things on behaviour of global cache here
1. Is the data ever evicted from global cache?
2. If my application is running high number of unique queries won't this cause the PgBouncer memory to keep increasing?

Thanks again for the help.

**Reply by JelteF** [reply]
1. Yes, when a query is not referenced by any client or server it is removed from the global cache.
2. Yes, it will continue to increase in that case, but only if you never close your client connections. Regularly closing of client connections should put a limit on the increase, e.g. closing a connection after 10 minutes. Usually client side connection poolers have settings to do this.
